SPECS/zstd/zstd.spec

```diff
@@ -1,96 +1,323 @@
+URL:            https://github.com/facebook/zstd
 Source0:        https://github.com/facebook/zstd/releases/download/v%{version}/%{name}-%{version}.tar.gz
-Requires:       %{name}-libs = %{version}-%{release}
+
+Patch1:         pzstd.1.patch
+
+BuildRequires:  make
+BuildRequires:  gcc %{?with_gtest:gtest-devel}
+%if %{with lz4}
+BuildRequires:  lz4-devel
+%endif
+%if %{with lzma}
+BuildRequires:  xz-devel
+%endif
+%if %{with pzstd}
+BuildRequires:  gcc-c++
+%endif
+%if %{with zlib}
+BuildRequires:  zlib-devel
+%endif
+BuildRequires:  execstack
 
 %description
-The zstd package contains programs for compressing and decompressing files
+Zstd, short for Zstandard, is a fast lossless compression algorithm,
+targeting real-time compression scenarios at zlib-level compression ratio.
 
-%package        devel
-Summary:        Header and development files for zstd
-Requires:       %{name} = %{version}-%{release}
-Provides:       lib%{name}-devel = %{version}-%{release}
+%package -n lib%{name}
+Summary:        Zstd shared library
 
-%description    devel
-Header and development files for zstd compression
+# For backwards compatibility
+Provides:       zstd-libs = %{version}-%{release}
+Obsoletes:      zstd-libs <= 1.5.5-5
 
-%package        libs
-Summary:        Libraries for zstd
-Group:          System Environment/Libraries
-Provides:       lib%{name} = %{version}-%{release}
+%description -n lib%{name}
+Zstandard compression shared library.
 
-%description    libs
-This package contains minimal set of shared zstd libraries.
+%package -n lib%{name}-devel
+Summary:        Header files for Zstd library
+Requires:       lib%{name}%{?_isa} = %{version}-%{release}
 
-%package        doc
-Summary:        Documentation files for zstd
-Requires:       %{name} = %{version}-%{release}
+# For backwards compatibility
+Provides:       zstd-devel = %{version}-%{release}
+Obsoletes:      zstd-devel <= 1.5.5-5
 
-%description    doc
-Documentation files for zstd
+%package -n lib%{name}-static
+Summary:        Static variant of the Zstd library
+Requires:       lib%{name}-devel = %{version}-%{release}
+
+%description -n lib%{name}-devel
+Header files for Zstd library.
+
+%description -n lib%{name}-static
+Static variant of the Zstd library.
 
 %prep
-%autosetup
+%setup -q
+find -name .gitignore -delete
+%if %{with pzstd}
+%patch1 -p1
+%endif
 
 %build
-%make_build
+export CFLAGS="$RPM_OPT_FLAGS"
+export LDFLAGS="$RPM_LD_FLAGS"
+export PREFIX="%{_prefix}"
+export LIBDIR="%{_libdir}"
+%make_build -C lib lib-mt %{!?with_asm:ZSTD_NO_ASM=1}
+%make_build -C programs %{!?with_asm:ZSTD_NO_ASM=1}
+%if %{with pzstd}
+export CXXFLAGS="$RPM_OPT_FLAGS"
 %make_build -C contrib/pzstd %{!?with_asm:ZSTD_NO_ASM=1}
+%endif
+
+%check
+execstack lib/libzstd.so.1
+
+export CFLAGS="$RPM_OPT_FLAGS"
+export LDFLAGS="$RPM_LD_FLAGS"
+make -C tests test-zstd
+%if %{with pzstd} && %{with gtest}
+export CXXFLAGS="$RPM_OPT_FLAGS"
+make -C contrib/pzstd test
+%endif
```

Comment:
Thanks for expanding the tests. This section may need a bit of rework, so we don't accidentally skip over failing tests.

More context:

The way a test pass/failure is currently figured out is from the `====== CHECK DONE rubygem-bindata. EXIT STATUS X` line RPM puts in the logs. The `X` is determined from the exit code of the last run command, so with mult-command `%check` sections we need to keep track of the exit code of at least the commands, which run the tests. One example would be inside `edk2.spec` where we set `tests_ok=true` at the beginning of the `%check` section and then update it to `false` whenever any of the tests fail. Then we make sure running `$tests_ok` is the last command of the `%check` section to capture the test failure if any of the subtests failed.

There are other ways to do it as well, but the general ask is make sure the `%check` section returns with a non-zero exit code when at least one of its tests fail.

Reply:
I did not expand the tests; this is directly from our inherited spec file. 

I do not follow what you are talking about in the %check section; if a test fails, it will return nonzero and fail the rpm build. There is no need to keep track of individual tests because the build is immediately failed on the first failure. There is nothing unusual with this %check section which, again, is directly from our inherited spec file and, as far as i can tell, is perfectly fine.

Reply:
Failing `%check` sections don't fail a package build as long as all other sections ran successfully. A shorter version of what I had in mind is: package test's status is determined by the exit code of the last command.

Reply:
Wait, nevermind, I now saw the latest change. Sorry for the noise.